wandb/sdk/launch/agent/agent.py

```diff
@@ -36,6 +39,97 @@
 _logger = logging.getLogger(__name__)
 
 
+def init_pool_processes(jobs: Dict[str, Dict[str, str]], lock: threading.Lock) -> None:
+    global _jobs
+    global _jobs_lock
+    _jobs = jobs  # type: ignore
+    _jobs_lock = lock  # type: ignore
+
+
+def thread_run_job(
+    launch_spec: Dict[str, Any],
+    job: Dict[str, Any],
+    default_config: Dict[str, Any],
+    api: Api,
+    jobs_event: threading.Event,
+) -> None:
+    try:
+        _run_job(launch_spec, job, default_config, api, jobs_event)
+    except Exception:
+        wandb.termerror(f"Error running job: {traceback.format_exc()}")
+        api.ack_run_queue_item(job["runQueueItemId"])
+
+
+def _run_job(
+    launch_spec: Dict[str, Any],
+    job: Dict[str, Any],
+    default_config: Dict[str, Any],
+    api: Api,
+    jobs_event: threading.Event,
+) -> None:
+    project = create_project_from_spec(launch_spec, api)
+    _logger.info("Fetching and validating project...")
+    project = fetch_and_validate_project(project, api)
+    _logger.info("Fetching resource...")
+    resource = launch_spec.get("resource") or "local-container"
+    backend_config: Dict[str, Any] = {
+        PROJECT_SYNCHRONOUS: False,  # agent always runs async
+    }
+
+    backend_config["runQueueItemId"] = job["runQueueItemId"]
+    _logger.info("Loading backend")
+    override_build_config = launch_spec.get("build")
+    override_registry_config = launch_spec.get("registry")
+
+    build_config, registry_config = resolve_build_and_registry_config(
+        default_config, override_build_config, override_registry_config
+    )
+    builder = load_builder(build_config)
+
+    default_runner = default_config.get("runner", {}).get("type")
+    if default_runner == resource:
+        backend_config["runner"] = default_config.get("runner")
+    backend = load_backend(resource, api, backend_config)
+    backend.verify()
+    _logger.info("Backend loaded...")
+    run = backend.run(project, builder, registry_config)
+
+    if not run:
+        return
+    with _jobs_lock:  # type: ignore
+        _jobs[run.id] = {"status": None}  # type: ignore
+    while jobs_event.is_set():
+        if _is_run_finished(run):
+            return
+        time.sleep(AGENT_POLLING_INTERVAL)
+    # temp: for local, kill all jobs. we don't yet have good handling for different
+    # types of runners in general
+    if isinstance(run, LocalSubmittedRun):
+        run.command_proc.kill()
+
+
+def _is_run_finished(run: AbstractRun) -> bool:
+    """Check our status enum."""
+    try:
+        status = run.get_status().state
+        with _jobs_lock:  # type: ignore
+            _jobs[run.id]["status"] = status  # type: ignore
+        if status in ["failed", "finished"]:
+            wandb.termlog(f"Job finished with ID: {run.id}")
+            return True
+        return False
+    except Exception as e:
```

Comment:
Maybe catch the `LaunchError` explicitly, catch a general Exception after, then do a `finally` with the logging stack trace?